nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -2007,6 +2057,265 @@ impl DataStore {
+
+            #[error("Serde error during Volume snapshot replacement: {0}")]
+            SerdeError(#[from] serde_json::Error),
+
+            #[error("Target Volume deleted")]
+            TargetVolumeDeleted,
+
+            #[error("Snapshot replacement error: {0}")]
+            SnapshotReplacementError(#[from] anyhow::Error),
+
+            #[error("Replaced {0} targets, expected {1}")]
+            UnexpectedReplacedTargets(usize, usize),
+
+            #[error("Updated {0} database rows, expected {1}")]
+            UnexpectedDatabaseUpdate(usize, usize),
+        }
+        let err = OptionalError::new();
+
+        let conn = self.pool_connection_unauthorized().await?;
+        self.transaction_retry_wrapper("volume_replace_snapshot")
+            .transaction(&conn, |conn| {
+                let err = err.clone();
+                async move {
+                    use db::schema::volume::dsl as volume_dsl;
+
+                    // Grab the old volume first
+                    let maybe_old_volume = {
+                        volume_dsl::volume
+                            .filter(volume_dsl::id.eq(volume_id))
+                            .select(Volume::as_select())
+                            .first_async::<Volume>(&conn)
+                            .await
+                            .optional()
+                            .map_err(|e| {
+                                err.bail_retryable_or_else(e, |e| {
+                                    VolumeReplaceSnapshotError::Public(
+                                        public_error_from_diesel(
+                                            e,
+                                            ErrorHandler::Server,
+                                        )
+                                    )
+                                })
+                            })?
+                    };
+
+                    let old_volume = if let Some(old_volume) = maybe_old_volume {
+                        old_volume
+                    } else {
+                        // Existing volume was deleted, so return an error. We
+                        // can't perform the snapshot replacement now!
+                        return Err(err.bail(
+                            VolumeReplaceSnapshotError::TargetVolumeDeleted
+                        ));
+                    };
+
+                    let old_vcr: VolumeConstructionRequest =
+                        match serde_json::from_str(&old_volume.data()) {
+                            Ok(vcr) => vcr,
+                            Err(e) => {
+                                return Err(err.bail(
+                                    VolumeReplaceSnapshotError::SerdeError(e)
+                                ));
+                            },
+                        };
+
+                    // Does it look like this replacement already happened?
+                    let old_target_in_vcr = match target_in_vcr(&old_vcr, &existing) {
```

Comment:
This will match either sub_volume or read_only_parent targets.
Is that okay?
The functions comment does state that it's only for replacing a read-only snapshot.

Reply:
addressed in 18837e338